packages/libs/deer/src/schema.rs

```diff
@@ -0,0 +1,641 @@
+
+    pub(crate) struct ArraySchema;
+    impl Reflection for ArraySchema {
+        fn schema(_: &mut Document) -> Schema {
+            Schema::new("array")
+        }
+    }
+
+    pub(crate) struct ObjectSchema;
+    impl Reflection for ObjectSchema {
+        fn schema(_: &mut Document) -> Schema {
+            Schema::new("object")
+        }
+    }
+
+    pub(crate) struct I8Schema;
+    impl Reflection for I8Schema {
+        fn schema(_: &mut Document) -> Schema {
+            Schema::new("integer")
+                .with("minimum", i8::MIN)
+                .with("maximum", i8::MAX)
+        }
+    }
+
+    pub(crate) struct I16Schema;
+    impl Reflection for I16Schema {
+        fn schema(_: &mut Document) -> Schema {
+            Schema::new("integer")
+                .with("minimum", i16::MIN)
+                .with("maximum", i16::MAX)
+        }
+    }
+
+    pub(crate) struct I32Schema;
+    impl Reflection for I32Schema {
+        fn schema(_: &mut Document) -> Schema {
+            Schema::new("integer")
+                .with("minimum", i32::MIN)
+                .with("maximum", i32::MAX)
+        }
+    }
+
+    pub(crate) struct I64Schema;
+    impl Reflection for I64Schema {
+        fn schema(_: &mut Document) -> Schema {
+            Schema::new("integer")
+                .with("minimum", i64::MIN)
+                .with("maximum", i64::MAX)
+        }
+    }
+
+    pub(crate) struct I128Schema;
+    impl Reflection for I128Schema {
+        fn schema(_: &mut Document) -> Schema {
+            Schema::new("integer")
+                .with("minimum", i128::MIN)
+                .with("maximum", i128::MAX)
+        }
+    }
+
+    pub(crate) struct ISizeSchema;
+    impl Reflection for ISizeSchema {
+        fn schema(_: &mut Document) -> Schema {
+            Schema::new("integer")
+                .with("minimum", isize::MIN)
+                .with("maximum", isize::MAX)
+        }
+    }
+
+    pub(crate) struct U8Schema;
+    impl Reflection for U8Schema {
+        fn schema(_: &mut Document) -> Schema {
+            Schema::new("integer")
+                .with("minimum", u8::MIN)
+                .with("maximum", u8::MAX)
+        }
+    }
+
+    pub(crate) struct U16Schema;
+    impl Reflection for U16Schema {
+        fn schema(_: &mut Document) -> Schema {
+            Schema::new("integer")
+                .with("minimum", u16::MIN)
+                .with("maximum", u16::MAX)
+        }
+    }
+
+    pub(crate) struct U32Schema;
+    impl Reflection for U32Schema {
+        fn schema(_: &mut Document) -> Schema {
+            Schema::new("integer")
+                .with("minimum", u32::MIN)
+                .with("maximum", u32::MAX)
+        }
+    }
+
+    pub(crate) struct U64Schema;
+    impl Reflection for U64Schema {
+        fn schema(_: &mut Document) -> Schema {
+            Schema::new("integer")
+                .with("minimum", u64::MIN)
+                .with("maximum", u64::MAX)
+        }
```

Comment:
I believe this is only a limitation of JS. The JSON RFC makes no distinction between number vs. integer.

> This specification allows implementations to set limits on the range and precision of numbers accepted [^1]

> Note that when such software is used, numbers that are integers and are in the range [-(2**53)+1, (2**53)-1] are interoperable in the sense that implementations will agree exactly on their numeric values. [^1]

As far as I understood, this means that implementations should implement 54-bit wide integers, but supporting larger ones is okay. For example, `serde_json` supports values larger than $2^{53} – 1$

[^1]: https://www.rfc-editor.org/rfc/rfc8259#section-6

Reply:
I see. It would be potentially incompatible then in those cases. I think it's fair to keep as-is for now :)
Perhaps it would be a good warning to provide that some implementations would not support the full value range of u64 if the built-in types above this number of bits is used